### PR TITLE
Test: Implemented ginkgo test against local kubeconfig

### PR DESCRIPTION
**Pull request by ellistarn**
Fixes # 

**Description**
Enables e2e tests against a local cluster. 
- test resources are all tagged with a special label for cleanup
- Made progress on CI infra setup (critical addons only, role name, etc), but gaps remain

```
 ~/w/g/s/g/aws/karpenter │ on testinfra !1  make e2etests                                2 ✘ │ at 18:00:08
ginkgo -r test -- --environment-name=karpenter-test-infrastructure
Running Suite: Integration Suite
================================
Random Seed: 1656550823
Will run 1 of 1 specs

• [SLOW TEST:85.879 seconds]
Sanity Checks
/Users/etarn/workspaces/go/src/github.com/aws/karpenter/test/suites/integration/suite_test.go:30
  should provision nodes
  /Users/etarn/workspaces/go/src/github.com/aws/karpenter/test/suites/integration/suite_test.go:31
------------------------------

Ran 1 of 1 Specs in 87.084 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 0 Skipped
PASS

Ginkgo ran 1 suite in 1m33.141041379s
Test Suite Passed
```

**How was this change tested?**

*

**Does this change impact docs?**
- [ ] Yes, PR includes docs updates
- [ ] Yes, issue opened: # 
- [ ] No

**Release Note**

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: test/pkg/environment/expectations.go
Comment (tzneal) [inline test/pkg/environment/expectations.go:34]: Shouldn't this merge the environment label instead of replacing existing?

File: test/pkg/environment/expectations.go
Comment (tzneal) [inline test/pkg/environment/expectations.go:47]: Maybe add `&client.DeleteAllOfOptions{DeleteOptions: client.DeleteOptions{GracePeriodSeconds: ptr.Int64(0)}})`

File: test/pkg/environment/expectations.go
Comment (tzneal) [inline test/pkg/environment/expectations.go:66]: Make this non-exported and same with `expectPodsHealthy`?

File: pkg/test/provisioner.go
Comment (tzneal) [inline pkg/test/provisioner.go:81]: Shouldn't fail, but wouldn't hurt to `Expect(err).To(BeNil())`

File: test/pkg/environment/environment.go
Comment (tzneal) [inline test/pkg/environment/environment.go:44]: Nice! Didn't know this existed. Maybe do `g.SetDefaultEventuallyPollingInterval(1*time.Second)` as well, looks like the default is 10ms

File: test/pkg/environment/expectations.go
Comment (ellistarn) [reply test/pkg/environment/expectations.go:47]: I wanted this to be closer to natural usage of clusters and include normal deletion flows rather than force deletion.

File: pkg/test/provisioner.go
Comment (ellistarn) [reply pkg/test/provisioner.go:81]: Trying to keep this decoupled from gomega. I'll panic instead.

Comment (ellistarn) [reply pkg/test/provisioner.go:81]: We're also totally ignoring the validation 😭 
`utilruntime.Must(provisioner.Validate(context.Background()))`

File: test/tasks/before-karpenter-suite.yaml
Comment (njtran) [inline test/tasks/before-karpenter-suite.yaml:46]: This is going to collide with our karpenter installation if using the same cluster name as the management cluster. Can we make it the following? It may be better to change up the host-cluster setup too so that we don't run into the chance of IAM Role name conflicts.
```suggestion
          --stack-name "KarpenterTesting-$(params.cluster-name)" \
```

File: test/tasks/go-suite.yaml
Comment (njtran) [inline test/tasks/go-suite.yaml:11]: Can you put descriptions for these

**Review by njtran — COMMENTED** [review]
Let’s discuss how we want to separate tasks, clusterTasks, and pipelines in the filesystem. If someone’s importing all the files from GitHub this is quite easy and nice to find the file definitions in the repo as you've done it. 

As the amount of configs increase: if someone’s doing this manually, they have to figure out how to find all related tasks then apply them manually. I think we should keep ClusterTasks in their own ClusterTasks folder and then any tasks in their respective suites. WDYT?

File: test/infrastructure/scripts/step-02-eksctl-cluster.sh
Comment (ellistarn) [reply test/infrastructure/scripts/step-02-eksctl-cluster.sh:18]: Can we do this separately? I didn't change he size of the nodegroup, just changed the max, which was strangely set to a high value. The only reason to have a max this high is if you're doing some sort of autoscaling on this group (which we're not). If we decide to manually scale this group in the UI, you can easily change max and desired. If we decide to change the size of this group in IAC, then the change should live in IAC.

Put simply, every value you configure has intent, and more apparent you can make that intent, the better. Having maxsize != desiredsize implies that we're going to autoscale this group -- which we aren't.

Comment (njtran) [reply test/infrastructure/scripts/step-02-eksctl-cluster.sh:18]: Sure. I think Brandon addressed this in his PR. Happy to defer to his, and I agree with the != maxsize point.

File: test/infrastructure/scripts/step-02-eksctl-cluster.sh
Comment (ellistarn) [reply test/infrastructure/scripts/step-02-eksctl-cluster.sh:1]: I have a bash linter/formatter in vscode that's cleaning this up. Do you have one with different opinions?

File: test/infrastructure/scripts/step-06-karpenter.sh
Comment (ellistarn) [reply test/infrastructure/scripts/step-06-karpenter.sh:28]: This is a really bad precedent to set. We can't be behind on karpenter in karpenter testing. Can't we set a higher ttl?

Comment (ellistarn) [reply test/infrastructure/scripts/step-06-karpenter.sh:28]: > Vending the Karpenter installation for KIT can be possible in the future and where they would establish the Karpenter version we could use.

Is this something like kitctl?

File: test/infrastructure/scripts/step-07-provisioners.sh
Comment (ellistarn) [reply test/infrastructure/scripts/step-07-provisioners.sh:18]: Good point.

File: test/infrastructure/scripts/step-02-eksctl-cluster.sh
Comment (njtran) [reply test/infrastructure/scripts/step-02-eksctl-cluster.sh:1]: Nope. Looks like I gotta get one for myself :)

File: test/infrastructure/scripts/step-07-provisioners.sh
Comment (ellistarn) [reply test/infrastructure/scripts/step-07-provisioners.sh:43]: How did I miss this.

File: test/infrastructure/scripts/step-06-karpenter.sh
Comment (njtran) [reply test/infrastructure/scripts/step-06-karpenter.sh:28]: Possibly. If this is an issue with pre-binding then I want to be able to figure this out. This was a fairly large change in assumptions for Karpenter so I don't think we'll be sticking here for long. We use a TTL of 180 right now.

File: test/infrastructure/scripts/step-02-eksctl-cluster.sh
Comment (bwagner5) [reply test/infrastructure/scripts/step-02-eksctl-cluster.sh:18]: Generally, you want maxSize to be 2x desired for upgrades. I'm not sure if MNG raises that for you when it does an upgrade?

File: test/suites/integration/pipeline.yaml
Comment (ellistarn) [reply test/suites/integration/pipeline.yaml:8]: I think it's self documenting as is.

File: test/suites/integration/pipeline.yaml
Comment (ellistarn) [reply test/suites/integration/pipeline.yaml:12]: Not a big fan of self naming objects. `deployments/foo-deployment` stutters

File: pkg/test/pods.go
Comment (ellistarn) [reply pkg/test/pods.go:96]: Pause doesn't work as an init container. Needed something that could exit.

File: pkg/test/provisioner.go
Comment (ellistarn) [reply pkg/test/provisioner.go:75]: Eventually sure -- I wanted to default for 10 until we need to override.

File: pkg/test/provisioner.go
Comment (ellistarn) [reply pkg/test/provisioner.go:92]: A ton of existing tests break if I fail here, since the error was originally squashed incorrectly. I wanted to log and clean it up later, since I'm OOTO for a few days and wanted to unblock folks.

File: test/infrastructure/scripts/step-01-config.sh
Comment (ellistarn) [reply test/infrastructure/scripts/step-01-config.sh:1]: `karpenter-test-management` is less idiomatic than `karpenter-test-infrastructure` to me.

**Review by njtran — CHANGES_REQUESTED** [review]
I think we have some disagreements on what the CI Infra right now. I think the ginkgo work looks great and am happy to approve that if we scope this one down.

File: test/pkg/environment/expectations.go
Comment (njtran) [inline test/pkg/environment/expectations.go:20]: How do we ensure that Nodes are cleaned up? I think the TTL on the Provisioners is meant to do that, but if a test fails to delete nodes (which could happen), then we want our Cleanup to also include those.

File: test/suites/integration/suite_test.go
Comment (njtran) [inline test/suites/integration/suite_test.go:32]: If the Karpenter being tested here doesn't have a defaultInstanceProfile, this should fail. I think it's fine to leave as is, but I think we should make sure we can test both modes eventually.

**Review by njtran — COMMENTED** [review]
Nice! Just a nit or two

File: test/pkg/environment/expectations.go
Comment (njtran) [reply test/pkg/environment/expectations.go:20]: Ah I forgot that Nodes are owned by provisioners now, so this should be good.

Comment (ellistarn) [reply test/pkg/environment/expectations.go:20]: We definitely need to test scale down, too. I think this library will need to be greatly expanded upon (e.g. wait for things to cleanup). This is just to get us started.

**Review by njtran — APPROVED** [review]

**Pull request merged**
